File: Artificial Intelligence/c++/DFS.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main() {
    int n = 6;

    vector<vector<int>> graph = {
        {1, 2},
        {0, 3, 4},
        {0, 4},
        {1, 5},
        {1, 2, 5},
        {3, 4}
    };

    vector<bool> visited(n, false);
    stack<int> st;

    st.push(0);

    cout << "DFS Traversal: ";

    while (!st.empty()) {
        int node = st.top();
        st.pop();

        if (!visited[node]) {
            visited[node] = true;
            cout << node << " ";

            for (int i = graph[node].size() - 1; i >= 0; i--) {
                int neighbor = graph[node][i];
                if (!visited[neighbor]) {
                    st.push(neighbor);
                }
            }
        }
    }

    return 0;
}

File: AI/Ex-Qn/G-II Solve/dfs.py
graph = {
    'A': ['B', 'C', 'D'],
    'B': ['E', 'F'],
    'C': [],
    'D': ['G'],
    'E': ['H'],
    'F': [],
    'G': [],
    'H': []
}

def dfs(graph, start, goal):
    visited = set()
    stack = [[start]]

    print("DFS Traversal Order:")
    while stack:
        path = stack.pop()
        node = path[-1]

        if node not in visited:
            print(node, end=' ')
            visited.add(node)

            if node == goal:
                print(f"\n\nGoal {goal} found!")
                print(f"Path: {' → '.join(path)}")
                return path

            for neighbor in reversed(graph[node]):
                if neighbor not in visited:
                    new_path = list(path)
                    new_path.append(neighbor)
                    stack.append(new_path)

    print(f"\nGoal {goal} not reachable")
    return None

dfs(graph, 'A', 'F')
